[donggi/Jay] 로또 3단계 - 수동구매 기능 추가

jinan159 opened the pull request:
리뷰어님 안녕하세요!  `donggi` , `Jay` 입니다.
**로또 3단계 - 수동구매 기능 추가** 미션 완료하여 리뷰 요청 드립니다.
항상 감사드립니다!!😊

# 3단계 간단 회고

- `donggi`
    기존 코드를 리펙토링을 하면서 객체의 역할과 책임에 대해 고민해볼 수 있는 시간이 되어 좋았습니다.
    책에서 봤던 내용을 머리속으로만 이해했었는데, 리펙토링을 하면서 책의 내용이 적용되는것을 확인할 수 있어서 좋았습니다.
    시간이 지날수록 합이 더 잘맞아가는것 같아서 좋았습니다.
- `Jay`
    기존에는 미션 요구사항 구현하는데 더 집중을 하느라 객체의 역할과 책임을 거의 신경을 쓰지 못했었고,
    객체간 협력 관계 없이 특정 부분에 너무 많은 로직을 넣었었습니다.
    그런데 동료와 미션을 진행하며, 막바지가 되서 본격적으로 리펙토링을 하면서 최대한 객체지향적인 구조를 만들기 위해 토론하는 것이
    정말 재밌었습니다. 제가 아는 내용을 공유하면서 무의식적으로 이해하고 있던 내용이 명확하게 정리가 되었고,
    동료가 이를 듣고 도움을 많이 받았다고 하셔서 뿌듯했습니다.
    그라운드룰을 정하고 이를 준수하려고 노력한 것이, 이번주에 지치지 않고 꾸준한 페어프로그래밍을 할 수 있었던 요인이라고 생각합니다.

# 이전단계 PR 반영사항(#26)

- [x] winningNumber + bonusNumber를 합쳐서 새로운 클래스 만들기
- [x] 하는 일에 비해 너무 많은 매개변수를 전달하는 로직 개선
- [x] LottoGame 클래스가 가져야할 멤버변수 고민

# 기능요구사항

- [x] 현재 로또 생성기는 자동 생성 기능만 제공한다. 사용자가 수동으로 추첨 번호를 입력할 수 있도록 해야 한다.
- [x] 입력한 금액, 자동 생성 숫자, 수동 생성 번호를 입력하도록 해야 한다.

# 프로그래밍 요구사항

- [x] 예외가 발생하는 부분에 대해 자바 Exception을 적용해 예외처리한다.
- [x] 사용자가 입력한 값에 대한 예외 처리를 철저히 한다.
- [ ] 상속과 인터페이스를 통해 구현을 간결히 할 수 없는지 고민해 본다.

# 고려한 사항

## 1. 로또 게임 전체 흐름을 단계별로 분리

로또 3단계 - 수동구매 기능 추가의 진행 흐름은 Main 클래스에 정의되어 있습니다.
처음에는 이 진행 흐름은 서로 유기적으로 연결되어있어서 분리가 힘들 것이라고 생각했습니다.
하지만, 자세히 살펴보다보니 분리하여 생각할 수 있는 것 같아 3단계로 분리하여 구현했습니다.

1. 고객 주문을 받고, 주문 결과를 보여주는 단계
    담당 객체 : `LottoCashier`
4. 당첨 번호 및 보너스 번호를 정하는 단계
    담당 객체 : `WinningLottoNumbers`
5. 당첨 결과를 출력하는 단계
     담당 객체 : `LottoGame`

이렇게 3단계로 분리하였고, 각 단계를 맡는 각각의 객체에게 역할과 책임을 부여하였습니다.
그래서 Main 에서는 전체 흐름만을 관리하고, 각 객체들은 자신이 맡은 단계에 대해서만 책임을 가지게 됩니다.

## 2. 게임 세부 로직의 추상화

위 내용에서 이어지는 내용입니다.
각 단계에서는 사실 많은 작업들이 일어나지만, 그 단계를 외부에서 볼때 한눈에 알아볼 수 있도록 추상화 하였습니다.
예를 들어 1단계의 경우, 구입금액, 수동 로또 개수를 입력받고 검증하여 자동 로또와 수동 로또를 개수에 맞게 생성합니다.

이런 로직들이 전체 흐름만 관리해야하는 Main에 노출된다면, 역할과 책임이 많아지게 되며 다른 로직으로 시야가 분산될 수 있습니다.
그래서 가능한 작업들을 최소화 하고, 외부에서 봤을때 유추 가능한 이름으로 작업의 이름(메서드명 등)을 지으려고 고민했습니다.

jinan159 committed:
수동으로 구매할 로또 개수 입력 기능 추가

jinan159 committed:
로또 수동구매 기능 추가

수동 구매한 로또를 출력하려다 보니, 불가피하게 기존 로직의 리펙토링도 함게 진행하였습니다.
- InputView 에 수동 구매 횟수와 수동 로또 번호 입력받는 로직 추가
- LottoController 의 getAllShuffledNumbers 메서드가 금액 대신 횟수를 인자로 받도록 변경하였습니다.
- OutputView 에 수동과 자동의 개수를 각각 출력하도록 변경하였습니다.
- LottoTicket 을 생성하는 팩토리 메서드를 추가하였습니다.

donggi-lee-bit committed:
수동으로 생성된 로또 번호 반영되지 않는 현상 개선 및 로또 게임 결과 출력 로직 리펙토링

- 로또 당첨 통계에 수동으로 생성된 로또 번호가 반영되지 않는 현상을 수정하였습니다.
- Main 클래스의 purchaseLotto 메소드를 LottoController 로 이동하였습니다.
- 로또 게임 전체 결과를 출력하기 위해 다른 계층으로 넘길 데이터를 담는 LottoGameResult 클래스를 추가하였습니다.

jinan159 committed:
LottoResult 클래스의 패키지 변경

기존에는 OutputView 에서만 LottoResult 를 사용할 거라고 생각하여
view 패키지에 포함시켰지만, 개발하다보니 다른 Lotto 로직에서도 많이
사용하게되어 domain 패키지로 이동하였습니다.

jinan159 committed:
LottoController 의 명칭을 LottoAutoMachine 으로 변경

jinan159 committed:
티켓을 관리하는 LottoTicketManager 클래스 생성

- 총 구매금액과 수동 발행 개수를 받아서 수동으로 발행할 티켓 개수와 자동으로 발행할 티켓 개수를 저장합니다.
- LottoAutoMachine에게 자동 발행 개수만큼 로또번호 생성을 요청합니다.
- 수동 로또 번호 입력 요청을 main 메서드에서 LottoTicketManager 로 이동하였습니다.

donggi-lee-bit committed:
당첨 여부를 확인하는 PrizeCounter 클래스 생성

- LottoGame 에서 당첨 결과를 반환해주는 책임이 있어 당첨 여부를 확인하는 책임을 PrizeCounter 를 생성하여 위임해주었습니다.

jinan159 committed:
main 메서드 로직 리펙토링

- Main 클래스의 main 메서드가 로또 게임의 큰 흐름만 제어하도록
세부 로직을 메서드로 분리하였습니다.
- LottoGameResult 에서 따로 관리하던 수동 로또와 자동 로또를 함께 관리하도록 변경하였습니다.

donggi-lee-bit committed:
당첨 번호와 보너스볼을 담는 새로운 클래스 추가

- 당첨 번호와 보너스볼을 각각 변수로 사용하다가 WinningLottoNumbers 라는 클래스에서 보너스볼과 당첨 번호를 함께 관리하도록 변경하였습니다.

jinan159 committed:
InputView 의 readCustomLottoCount 메소드 오류 수정

- readCustomLottoCount 가 재입력이 필요할 경우 readAmount 를 재귀호출하는 오류를 수정하였습니다.
- 검증 메소드인 readAmount 의 예외 메세지가 1~45 로 고정된 오류를 수정하였습니다.

jinan159 committed:
중복 검증 로직 추가

- HashSet을 이용하여 기존의 로또 번호 배열의 길이와 HashSet의 길이를 비교하여 다르면 중복된 번호가 있다는 예외가 발생합니다.

jinan159 committed:
InputViewValidator 중복된 로직을 메소드로 분리

String 공백을 제거하여 int 로 변환하는 중복된 로직을 메소드로 분리하였습니다.

jinan159 committed:
LottoTicket 이 일급컬렉션의 조건을 만족하도록 리펙토링

- LottoTicket 생성시에만 데이터를 주입하고, 생성 후에는 조회만 가능하도록 수정하였습니다.

jinan159 committed:
불필요한 import 제거

jinan159 committed:
LottoGame 이 로또 게임 진행에 필요한 정보를 저장하도록 변경

- 게임 진행에 필요한 정보를 담는 LottoGameResult 를 LottoGameDto 로 명칭 변경

jinan159 committed:
수동, 자동 로또 티켓 생성을 LottoTicketManager 생성자에서 처리하도록 변경

donggi-lee-bit committed:
주문을 받아 LottoTicketManager 에게 로또 생성 요청을 하는 LottoCashier 클래스 추가

- Main 에서 로또 주문을 받는 책임을 LottoCashier 로 위임하였습니다.

jinan159 committed:
Main 클래스의 로또 당첨번호 입력 요청 책임을 WinningLottoNumbers 로 위임

jinan159 committed:
LottoGameDto 클래스 빌더패턴 제거

생성자 인자가 2개 뿐이라 빌더패턴을 사용하면 더 복잡하게 보여서, 제거하였습니다.

donggi-lee-bit committed:
로또 번호가 정렬되지 않던 오류 수정

donggi-lee-bit committed:
validateAmountUnit 메서드 에러 출력 메시지 수정

- 1000원 단위가 아닌 값을 넘겨받은 unit 단위로 입력해달라는 메시지로 수정하였습니다.

donggi-lee-bit committed:
InputViewValidator 테스트 코드 추가

jinan159 committed:
README 작성할 대제목 추가

jinan159 committed:
LottoCashier 에서 구매금액을 final 로 관리하도록 개선

donggi-lee-bit committed:
주요 로직에 대한 리드미 추가

jinan159 committed:
README 프로젝트 구조 및 고려한 사항 추가

tmdgusya commented on src/main/java/app/lotto/Main.java:
```diff
@@ -1,43 +1,20 @@
-
-        List<LottoTicket> allShuffledNumbers = LottoController.getAllShuffledNumbers(amount);
+        WinningLottoNumbers winningLottoNumbers = WinningLottoNumbers.readWinningNumbersAndCreate();
+        System.out.println();
```
ㅎㅎ 이건 단지 뛰어쓰기용인가요?

donggi-lee-bit replied:
네 맞습니다..!!

tmdgusya commented on src/main/java/app/lotto/domain/LottoAutoMachine.java:
```diff
@@ -0,0 +1,40 @@
+    private static LottoTicket getAutoLottoNumbers() {
+        List<Integer> lottoNumbers = getLottoNumbers();
+        Collections.shuffle(lottoNumbers);
+        return new LottoTicket(lottoNumbers.subList(0, 6));
```
0 과 6 은 어떤의미일까요? 매직넘버는 상수로 추출하시죠.

donggi-lee-bit replied:
매직넘버를 놓치고 있었습니다. 상수로 추출하겠습니다 감사합니다!

tmdgusya commented on src/main/java/app/lotto/domain/LottoAutoMachine.java:
```diff
@@ -0,0 +1,40 @@
+    }
+
+    private static List<Integer> getLottoNumbers() {
+        return IntStream.rangeClosed(1,45)
```
1 과 45 는 어떤 의미일까요? 이것또한 상수로 추출하시죠.

donggi-lee-bit replied:
이것 또한 상수로 추출하겠습니다!

tmdgusya commented on src/main/java/app/lotto/domain/LottoCashier.java:
```diff
@@ -0,0 +1,44 @@
+        this.lottoTicketManager = LottoTicketManager.createWithTotalAmountAndCustomTicketCount(amount, customLottoCount);
+    }
+
+    public static LottoCashier receiveOrderAndCreate() {
```
약간 문맥상 이해하기 힘들다고 생각하는데 한번 naming 과 return type 에 대해 생각해보시면 좋을거 같아요

`receiveOrderAndCreate` -> 주문을 받고 생성한다 정도로 의역을 했는데 그 ReturnType 이 LottoCashier 인게 이해가 안가네요. 문맥만 보면 로또 주문을 받고 로또를 생성해서 리턴해준다 같은데요 ㅎㅎ..

tmdgusya commented on src/main/java/app/lotto/domain/LottoGameDto.java:
```diff
@@ -0,0 +1,24 @@
+import java.util.Collections;
+import java.util.List;
+
+public class LottoGameDto {
```
Dto 를 쓰신이유는 무엇일까요?
그냥 단순히 궁금해서 물어봅니다 ㅎㅎ

jinan159 replied:
`DTO`를 만들게된 배경은, 기존에 역할과 책임이 제대로 분리되지 않은 로직에서
`LottoGame` 클래스의 특정 메소드에 너무 많은 파라미터(6~7개)를 넘기는 부분이 있었습니다.
그 많은 파라미터를 함께 전달하고자 `DTO`를 만들었었는데, 리펙토링을 하게되며 많은 부분이 개선되었습니다.
그래서 최종적으로 2개의 속성만을 가지는 DTO가 만들어지게 되었습니다.

기존에 많은 속성들(6~7개)이 있던 LottoGameDto 는 속성들이 서로 연관된것도 있고, 연관짓기 힘든것도 있었습니다.
그래서 딱 뭐라고 명명하기가 애매하여 `DTO`라는 명칭을 붙였습니다.
그런데 리펙토링 후 `로또 티켓들`과 `로또 당첨번호`만 남게되어, 더 명확한 이름으로 변경할 수 있었지만
다른 더 급한 수정사항이 많아서 잊고 지나가게 된것 같습니다.

tmdgusya commented on src/main/java/app/lotto/domain/LottoAutoMachine.java:
```diff
@@ -0,0 +1,40 @@
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+public class LottoAutoMachine {
```
클래스 전반적으로 `static` 인 부분이 많은데 굳이 모든 메소드가 static 이 여야 하나? 라는 생각이 있네요.
잘 생각해 보셔야 할거 같아요. 
제가 코드를 봤을때는 `static` 에 전염 당해 모든 메소드가 static 을 써야만 하는 상황처럼 느껴지기도 하네요.

tmdgusya replied:
보통 Util 성의 클래스들이 해당 클래스처럼 대부분의 method 가 static 으로 작성하는데요.
저는 LottoAutoMachine 이 과연 그런 클래스 일까? 라는 의문이 있습니다.
잘 생각해보시고 페어 두분께서 결정해보시죠~!

tmdgusya commented on src/main/java/app/lotto/domain/LottoGame.java:
```diff
@@ -1,79 +1,59 @@
+        return new LottoGame(lottoGameDto, amount);
+    }
+
+    public void printLottoGameResult() {
```
👍

tmdgusya commented on src/main/java/app/lotto/domain/LottoTicketManager.java:
```diff
@@ -0,0 +1,52 @@
+    }
+
+    public List<LottoTicket> getCustomLottoTickets() {
+        return Collections.unmodifiableList(customLottoTickets);
```
불변형 좋네요 👍

tmdgusya commented on src/main/java/app/lotto/domain/PrizeCounter.java:
```diff
@@ -0,0 +1,49 @@
+import java.util.List;
+import java.util.Map;
+
+public class PrizeCounter {
```
통계나 결과를 계산하는 클래스이군요.
좀더 좋은 네이밍을 고려해보시죠 ㅎㅎ

tmdgusya commented on src/main/java/app/lotto/domain/WinningLottoNumbers.java:
```diff
@@ -0,0 +1,26 @@
+
+import app.lotto.view.InputView;
+
+public class WinningLottoNumbers {
```
따로 빼신 부분 좋네요 ㅎㅎ

tmdgusya commented on src/test/java/app/lotto/validation/InputViewValidatorTest.java:
```diff
@@ -0,0 +1,198 @@
+
+import static org.junit.jupiter.api.Assertions.*;
+
+class InputViewValidatorTest {
```
테스트 깔끔하네요 ㅎㅎ 좋습니다 💯

tmdgusya reviewed (changes requested):
테스트를 작성하신 부분 좋았습니다.  코드도 전반적으로 깔끔하네요 💯 

아직 매직넘버를 상수로 추출하는 부분이 익숙하지는 않으신 것 같습니다. 앞으로 코드를 칠 때는 조금 더 매직넘버를 유의하여 코드로 작성해주세요.

전반적으로 static 을 많이 사용한다고 느껴지는데요.
어느 부분에서 static 으로 객체 메소드를 유지해야 할지, 아니면 그냥 instance 가 나은지 생각해보시는게 좋을거 같아요. 약간 무분별하게 사용되는 부분도 있는거 같아 보입니다.

LottoMaker 를 수동 / 자동으로 나누는 부분은 어렵지만 고려를 한번 해보시는게 좋아요.
다만 어렵다면 어쩔 수 없지요. 
코드에 녹이기 어렵더라도 **인터페이스 / 추상클래스 / 합성** 에 관한 부분은 따로 학습해 보시면 좋을 겁니다.

몇가지 부분 수정해서 올리시면 머지 해드리겠습니다 ㅎㅎ

donggi-lee-bit committed:
매직넘버 상수로 추출

- 1부터 45까지의 로또 생성 번호 범위를 상수로 추출하였습니다.
- 6개의 자동으로 생성된 로또 번호를 가져오기 위한 숫자 범위 또한 상수로 추출하였습니다.

donggi-lee-bit committed:
LottoAutoMachine 에서 금액에 따른 로또 개수를 계산하는 책임을 LottoCashier 로 위임

jinan159 committed:
LottoAutoMachine 을 객체로 변경하고, LottoTicketManager 의 속성으로 변경

객체간의 참조관계를 단순화하기 위해서 LottoTicketManager 만 LottoAutoMachine 을 참조하고
 외부에서는 LottoTicketManager createAutoLottoTickets 메소드만 알고있으면
 자동 생성된 번호를 가져올 수 있다고 알게하기 위해서 리펙토링 하였습니다.

jinan159 committed:
PrizeCounter 를 객체로 변경하며, 더 적절한 명칭으로 변경

- PrizeCounter 를 더 적절하게 표현할 수 있는 PrizeResultCalculator 로 명칭을 변경하였습니다.
- LottoGame 과 PrizeResultCalculator 가 협력하여 결과를 출력할 수 있도록,
 PrizeResultCalculator 를 static 형태에서 객체 형태로 변경하였습니다.

donggi-lee-bit committed:
LottoCashier 팩토리 메서드명 수정

- 팩토리 메서드의 의미가 잘 나타나지 않는 것 같아 의미가 더 드러나도록 수정하였습니다.

jinan159 commented:
## PR 반영사항

- `매직넘버`를 `상수`로 분리
- `static` 을 남용하던 클래스를, 객체로 사용되도록 변경
- `PrizeCounter` 클래스 명칭을 더 의미가 명확하게 전달되도록 `PrizeResultCalculator` 로 변경
- `LottoCashier` 클래스의 팩토리 메소드 명칭을 의미가 더 명확하게 전달되게 수정

## 고민한것들

### 0 같은 숫자를 상수로 추출?

`매직넘버`를 `상수`로 추출하면서 의미를 생각하다 보니, `0`과 같은 숫자들은 `상수`로 추출하니 오히려 무슨 의미인지 더 헷갈렸습니다.
 그래서 이 부분은 `0`으로 두고, 나머지만 `상수`로 추출했습니다.
`0`이 subList 라는 메소드와 함께 쓰이면, **첫 번째 인덱스라는 의미가 몀확하게 전달된다고 생각**하여 `상수`로 추출하지 않았는데,
이 부분에 대한 리뷰어님의 의견이 궁금합니다.

### 팩토리 메소드의 네이밍

기존에 팩토리 메소드 이름을 지을때, **주문을 받으면서 `LottoCashier` 라는 객체를 초기화** 하고 싶어서 `receiveOrderAndCreate`
라는 이름으로 지었었습니다.
하지만 제 3자의 시선(리뷰어님)으로 봤을때는, 그 의미가 명확하게 전달되지 않은 것 같았습니다.
그래서 팩토리 메소드의 이름만 변경하였는데, 변경된 이름이 의도한대로 의미가 전달되는지 궁금합니다.

tmdgusya commented:
질문 주신대로 0 의 경우 추출할때도 있고, 없을때도 있는데요.
생각해보니 LOTTO_MAX_SIZE (네이밍이 적절한지는 모르겠네요 ㅎㅎ;) 저런식으로 되어있었다면 안헷갈렸을거 같기도 하네요.
근데 상수로 추출하면 오히려 더 헷갈릴 이유는 없지 않을까요? 헷갈린다면 네이밍의 문제라고 생각합니다.

`receiveOrderAndCreate` 보다는 만드는 것이니 `createXXX` 식이 낫다고 생각했는데 잘 바꾸셨네요 ㅎㅎ 💯 
고생하셨습니다
